### PR TITLE
#10547: Change kernel to support concat width or height-sharded tensors

### PR DESCRIPTION
**Pull request by jerrysky3**
### Ticket
#10547

### Problem description
Add kernel to support concat on both width and height-sharded tensors. Currently it supports the following cases:
- Height-sharded width concat
- Width-sharded height concat (the new case supported by this change)

### What's changed
- Add a new kernel `reader_s2s_rm_tensor_concat.cpp` to support both concat cases
- Update the implementation of `s2s_rm_concat_multi_core` to support both concat cases
  - The approach is documented in the comments
  - It also replaces the `s2s_rm_concat_two_tensors_multi_core` for two tensors special case
- Validate supported sharding cases and grid config in `ConcatDeviceOperation::validate`
- Add tests for both cases and with/without program cache enabled

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp [hunk 1]
```diff
@@ -3,352 +3,156 @@
...
+// Concat sharded tensors into sharded output in row-major layout. Currently it only supports height-sharded width
+// concat or width-sharded height concat.
+//
+// It is done by copying each row of each input sharded tensor to the right offset in the sharded output tensor based on
+// the sharded output width in bytes (output stride). This way works for both width and height concat.
+//
+// For example in width concat, rows of an input tensor are placed at the same column offset but sequential rows in the
+// output. The memory address gap between neighbor input rows is exactly the output width. In height concat, all input
+// rows are placed at column 0 but sequential rows in the output. The address gap between neighbor input rows is still
+// the output width (which is equal to the input width).
```
Comment (jerrysky3) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:43]: For now I directly replaced the `s2s_rm_concat_multi_core` and removed `s2s_rm_concat_two_tensors_multi_core` for the two tensors as the new kernel should support the cases they covered.

Let me know if we want to put this new kernel as experimental separately and maybe guarded by a flag. Or if there is a special reason we want to keep `s2s_rm_concat_two_tensors_multi_core`

**Comment by jerrysky3** [comment]
This change has passed post-commit regression tests locally on n150. But I'd like to know how to check if this change causes any performance regressions?

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp (hunk 1 above)
Comment (ntarafdar) [reply ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:43]: ya two tensors multi core is just a manually unrolled version of the general case (it should be covered by it).

but we had a model where concat was the bottleneck and using runtime args (which is how we added variadic number of inputs) added a bunch of cycles. I think for now guarding against that with a flag is okay (if input is 2 tensors ) . 

I know @TT-billteng is looking for efficient ways to send variadic args in compile-time args and after that we should be able to remove the manually unrolled kernel.

**Comment by ntarafdar** [comment]
Overall this is amazing! good job @jerrysky3 . I think just the comment on leaving the manually unrolled kernel in for now. 
Good work!  After this week we will have a lot more issues for you guys to work on. It's great that you can take on the sharding example from the get go  as this is supposed to be a harder case.

**Review by ntarafdar — APPROVED** [review]

**Comment by ntarafdar** [comment]
> This change has passed post-commit regression tests locally on my machine with n150. But I'd like to know how to check if this change causes any performance regressions?

https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml
https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml
https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp (hunk 1 above)
Comment (jerrysky3) [reply ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:166]: SG. I restored the original two tensors concat kernel. If we want to also support width-sharded hight concat with two tensors in the meantime, I'm happy to open another PR to replace `s2s_rm_concat_two_tensors_multi_core` with an unrolled version of this new kernel

Comment (jerrysky3) [reply ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:166]: Regarding to using a flag to guard new kernel for the input is 2 tensors, I'm actually not clear about how the flag is used in tt-metal library. So for now I just restore the `s2s_rm_concat_two_tensors_multi_core` to keep it simple.

Let me know if we really want a flag and if there is an example on how the flag is used

**Comment by jerrysky3** [comment]
> > This change has passed post-commit regression tests locally on my machine with n150. But I'd like to know how to check if this change causes any performance regressions?
> 
> https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml

Thanks for the review and feedback. I'm trying to follow https://github.com/tenstorrent/tt-metal/blob/9cc51a329e685a6fdd228ecce8860f98fe9c66e9/CONTRIBUTING.md#using-cicd-for-development to run these workflows but it seems that the workflow run is only allowed for in-repo branches and TT team members (related to  #13383)

Hi @ayerofieiev-tt, I think this might be an issue for upcoming kernel development

**Comment by jerrysky3** [comment]
Hi @ntarafdar, in order to run CI, I opened a new PR https://github.com/tenstorrent/tt-metal/pull/13737 with in-repo branch and closed this one.

The new PR is identical to this one. Could you help take a look at the [new one](https://github.com/tenstorrent/tt-metal/pull/13737) and give an approval there if everything looks fine to you? Thanks!

**Pull request closed without merging**